packages/agreement-platformstate-writer/src/utils.ts

```diff
@@ -92,3 +111,434 @@ export const deleteAgreementEntry = async (
+    GSIPK_eserviceId_descriptorId: GSIPKEServiceIdDescriptorId;
+    catalogEntry: PlatformStatesCatalogEntry | undefined;
+  }): Promise<TokenGenerationStatesClientPurposeEntry[]> => {
+    const runPaginatedQuery = async (
+      consumerId_eserviceId: GSIPKConsumerIdEServiceId,
+      dynamoDBClient: DynamoDBClient,
+      exclusiveStartKey?: Record<string, AttributeValue>
+    ): Promise<TokenGenerationStatesClientPurposeEntry[]> => {
+      const input: QueryInput = {
+        TableName: config.tokenGenerationReadModelTableNameTokenGeneration,
+        IndexName: "Agreement",
+        KeyConditionExpression: `GSIPK_consumerId_eserviceId = :gsiValue`,
+        ExpressionAttributeValues: {
+          ":gsiValue": { S: consumerId_eserviceId },
+        },
+        ExclusiveStartKey: exclusiveStartKey,
+      };
+      const command = new QueryCommand(input);
+      const data: QueryCommandOutput = await dynamoDBClient.send(command);
+
+      if (!data.Items) {
+        throw genericInternalError(
+          `Unable to read token state entries: result ${JSON.stringify(data)} `
+        );
+      } else {
+        const unmarshalledItems = data.Items.map((item) => unmarshall(item));
+
+        const tokenStateEntries = z
+          .array(TokenGenerationStatesClientPurposeEntry)
+          .safeParse(unmarshalledItems);
+
+        if (!tokenStateEntries.success) {
+          throw genericInternalError(
+            `Unable to parse token state entry item: result ${JSON.stringify(
+              tokenStateEntries
+            )} - data ${JSON.stringify(data)} `
+          );
+        }
+
+        await updateAgreementStateInTokenGenerationStatesTableEntriesPlusDescriptorInfo(
+          {
+            entriesToUpdate: tokenStateEntries.data,
+            agreementState,
+            dynamoDBClient,
+            GSIPK_eserviceId_descriptorId,
+            catalogEntry,
+          }
+        );
+
+        if (!data.LastEvaluatedKey) {
+          return tokenStateEntries.data;
+        } else {
+          return [
+            ...tokenStateEntries.data,
+            ...(await runPaginatedQuery(
+              consumerId_eserviceId,
+              dynamoDBClient,
+              data.LastEvaluatedKey
+            )),
+          ];
+        }
+      }
+    };
+
+    return await runPaginatedQuery(
+      GSIPK_consumerId_eserviceId,
+      dynamoDBClient,
+      undefined
+    );
+  };
+
+export const extractAgreementIdFromAgreementPK = (
+  pk: PlatformStatesAgreementPK
+): AgreementId => {
+  const substrings = pk.split("#");
+  const agreementId = substrings[1];
+  const result = AgreementId.safeParse(agreementId);
+
+  if (!result.success) {
+    throw genericInternalError(
+      `Unable to parse agreement PK: result ${JSON.stringify(
+        result
+      )} - data ${JSON.stringify(agreementId)} `
+    );
+  }
+  return result.data;
+};
+
+export const updateAgreementStateInTokenGenerationStatesTable = async ({
```

Comment:
```suggestion
export const updateAgreementOnTokenStatesQuery = async ({
```

Reply:
Renamed to `updateAgreementStateOnTokenStates`